src/sys/fuchsia/selector.rs

```diff
@@ -0,0 +1,374 @@
+                                .saturating_add(duration.subsec_nanos() as u64);
+
+                magenta::deadline_after(nanos)
+            }
+            None => magenta::MX_TIME_INFINITE,
+        };
+
+        let packet = match self.port.wait(deadline) {
+            Ok(packet) => packet,
+            Err(magenta::Status::ErrTimedOut) => return Ok(false),
+            Err(e) => return Err(status_to_io_err(e)),
+        };
+
+        let observed_signals = match packet.contents() {
+            magenta::PacketContents::SignalOne(signal_packet) => {
+                signal_packet.observed()
+            }
+            magenta::PacketContents::SignalRep(signal_packet) => {
+                signal_packet.observed()
+            }
+            magenta::PacketContents::User(_user_packet) => {
+                // User packets are only ever sent by an Awakener
+                return Ok(true);
+            }
+        };
+
+        let key = packet.key();
+        let (token, reg_type) = token_and_type_from_key(key);
+
+        match reg_type {
+            RegType::Handle => {
+                // We can return immediately-- no lookup or registration necessary.
+                evts.events.push(Event::new(signals_to_ready(observed_signals), token));
+                Ok(false)
+            },
+            RegType::Fd => {
+                // Convert the signals to epoll events using __mxio_wait_end,
+                // and add to reregistration list if necessary.
+                let events: u32;
+                {
+                    let handle = if let Some(handle) =
+                    self.token_to_fd.lock().unwrap()
+                        .get(&token)
+                        .and_then(|h| h.upgrade()) {
+                        handle
+                    } else {
+                        // This handle is apparently in the process of removal.
+                        // It has been removed from the list, but port_cancel has not been called.
+                        return Ok(false);
+                    };
+
+                    events = unsafe {
+                        let mut events: u32 = mem::uninitialized();
+                        sys::fuchsia::sys::__mxio_wait_end(handle.mxio(), observed_signals, &mut events);
+                        events
+                    };
+
+                    // If necessary, queue to be reregistered before next port_await
+                    let needs_to_rereg = {
+                        let registration_lock = handle.registration().lock().unwrap();
+
+                        registration_lock
+                            .as_ref()
+                            .and_then(|r| r.rereg_signals())
+                            .is_some()
+                    };
+
+                    if needs_to_rereg {
+                        let mut tokens_to_rereg_lock = self.tokens_to_rereg.lock().unwrap();
+                        tokens_to_rereg_lock.push(token);
```

Comment:
Wait, I just realized I have no idea why the handle isn't just re-registered right here and is re-registered later instead.